### ISSUE TITLE
Custom Function all always trigger when click on every cell 

### ISSUE THREAD

**Issue by bunheng168** [odoo/o-spreadsheet#1858]
I want to query records from database via API to fill on cells by adding custom function and custom plugin to do it. 

similar to this on google spreadsheet
oe_browse("sale.order";"name date_order commitment_date expected_date partner_id client_order_ref website_id user_id team_id amount_untaxed amount_tax amount_total invoice_status state activity_exception_decoration";"[]")

Custom Function 
```
compute: function (a, b) {
this.getters.fillCell(this.getters.getActiveSheetId(),a,b)
return "test"
}
```
Plugin 
```
class QueryPlugin extends CorePlugin {
    static getters = ["fillCell"];
  
    constructor(getters, history, dispatch, config) {
      super(getters, history, dispatch, config);

    }
  
    fillCell(sheetId,a,b) {
        console.log(1)
        const col = a;
        const row = b;
        this.dispatch("UPDATE_CELL", {
            col,
            row,
            sheetId,
            content: "Hello world",
            format: "",
          });
        this.dispatch("EVALUATE_CELLS",{ sheetId: sheetId});
        this.isStale = true
        return { status: "COMPLETED" };
    }
}
  
corePluginRegistry.add("QueryPlugin", QueryPlugin);
```

**Comment by pro-odoo** [comment]
Hello,

Thanks for your interest in o-spreadsheet :) 

The reason why you have a loop is because you use `dispatch` in the getter that is called from the formula. After every `dispatch` of a command that could invalidate the evaluation [(1)](https://github.com/odoo/o-spreadsheet/blob/16.0/src/plugins/ui/evaluation.ts#L67), the evaluation process is triggered [(2)](https://github.com/odoo/o-spreadsheet/blob/16.0/src/plugins/ui/evaluation.ts#L78), and so your getter is re-called, re-dispatch a command, etc.

> compute: function (a, b) {
this.getters.fillCell(this.getters.getActiveSheetId(),a,b)
return "test"
}

From what I understand, you want here to change the `content` of a cell from another cell, which is not allowed in a spreadsheet. The formula can only change the value of its cell, or change the value of a **zone** if the result should expand (e.g. `TRANSPOSE` function).
The latter is not yet supported in o-spreadsheet, but will come at some point.

What you have to do is to store the result of your request in a cache, and have a getter that will returns this value if available, or trigger the request if the value is not available. You can find [here](https://github.com/odoo/o-spreadsheet/blob/16.0/doc/add_function.md#asynchronous-function) a link to something similar. In this exemple, we dispatch an `EVALUATE_CELLS` after the request is completed.

**Comment by bunheng168** [comment]
@pro-odoo, thanks for your information. How about update multiple cells or return array in custom function, it is possible ?

**Comment by pro-odoo** [comment]
Hello @bunheng168,

It's not yet supported, but it's something we're currently working on.

**Comment by bunheng168** [comment]
@pro-odoo  When you release?

**Comment by rrahir** [comment]
Hi, The pull request is open and in current review, you can  track its evolution here https://github.com/odoo/o-spreadsheet/pull/2090

**Comment by rrahir** [comment]
will close it once multicell is merged

**Comment by rrahir** [comment]
Good news!
The array formulas were released in #2090 
You can find a simple example of implementation here: https://github.com/odoo/o-spreadsheet/blob/17.0/src/functions/module_math.ts#L883
and more complex one inside Odoo here: https://github.com/odoo/odoo/blob/17.0/addons/spreadsheet/static/src/pivot/pivot_functions.js#L129

Thanks again for your interest in o-spreadsheet. DOn't hesitate to reopen the issue if you still have questions related to this subject.
